File: dotnet/Allmark/Render/RenderTag.cs
namespace Allmark.Render;

using Allmark.Types;

public static class RenderTag
{
    public static void Execute(MarkdownNode node, RendererState state, string tag, bool? decode = true)
    {
        RenderUtils.StartNewLine(node, state);
        state.Output.Append($"<{tag}>");

        if (node.Block && (node.Children?.Count ?? 0) == 0)
        {
            state.Output.Append("\n");
        }
        else
        {
            RenderUtils.InnerNewLine(node, state);
            RenderChildren.Execute(node, state, decode);
            if (node.Block)
            {
                if (state.Output.Length > 0 && state.Output[^1] == '\n')
                {
                    state.Output.Length -= 1;
                }
                if (state.Output.Length > 0 && state.Output[^1] == '\r')
                {
                    state.Output.Length -= 1;
                }
                state.Output.Append('\n');
            }
        }

        state.Output.Append($"</{tag}>");
        RenderUtils.EndNewLine(node, state);
    }
}

File: zig/src/render/renderTag.zig
const std = @import("std");
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const RendererState = @import("../types/RendererState.zig").RendererState;
const renderChildren = @import("renderChildren.zig").renderChildren;
const renderUtils = @import("renderUtils.zig");

pub fn renderTag(
    node: *const MarkdownNode,
    state: *RendererState,
    tag: []const u8,
    decode: bool,
) void {
    renderUtils.startNewLine(node, state);
    state.output.writer().print("<{s}>", .{tag}) catch unreachable;
    if (node.block) {
        if (node.children == null or node.children.?.len == 0) {
            state.output.append('\n') catch unreachable;
        } else {
            renderUtils.innerNewLine(node, state);
        }
    }
    renderChildren.renderChildren(node, state, decode);
    state.output.writer().print("</{s}>", .{tag}) catch unreachable;
    renderUtils.endNewLine(node, state);
}
